### PR TITLE
ALTAPPS-632: Android stage implementation

### PR DESCRIPTION
**Pull request by XanderZhu**
**YouTrack Issues**:
[#ALTAPPS-632](https://vyahhi.myjetbrains.com/youtrack/issue/ALTAPPS-632)

**Checklist**

_Before Code Review:_

- [x] Fields "Assignees, Labels, Milestone" are filled in the pull request;
- [x] All checks have been passed;
- [x] Changes have been checked locally.

**Description**

_StepQuiz ui refactoring_:
- Move stepQuiz title from `StepPractice` to the certain `StepQuiz` ui;
- Move codeSamples from `StepPractice` to the CodeStepQuizFragment to easily reuse `CodeStepQuizFragment` between `Step` and `Stage` views;

_StageImplementationFragment_
- Implement `StageStepWrapperFragment` to define custom view for `StageImplementation`. It uses `StepViewModel` and uses `StepQuizFragmentFactory` to launch certain stepQuiz fragment;
- Implement `StageImplementationFragment` as an entry point for stage implementation feature. It launches `StageStepWrapperFragment`;

_`Pycharm` step quiz support_:
- Refactor `CodeStepQuizFragment` with `CodeStepQuizConfig` to allow different languages usage;
- Support `Pycharm` stepQuiz using `PycharmCodeQuizConfig`;
- Implement `SqlCodeStepConfig` to support `CodeStepQuizFullScreenDialogFragment` (`SqlStepQuizFragment` is supposed to be replaced with just a `CodeStepQuizFragment` in future)
- Add `pycharmCode` extension for `Reply` and `Step` to parse `Pycharm` quiz code

_Unsupported state for stepQuiz:_
- Implement `Unsupported` state handling in `DefaultStepQuizFragment`;
- Move `StepQuizHints` from `StepPracticeFragment` to `DefaultStepQuizFragment` to launch it only when `StepQuizFeature.State == StepQuizFeature.State.Unsupported`
- Implement `FindStage` block in `DebugFragment` to access `StageImplementationFragment`

### CODE REVIEW

**Comment by ivan-magda** [comment]
When the submission status is wrong pycharm problem doesn't has a reset button (please also check all code-related problems SQL, code and pycharm):


 


Code reset works incorrectly for the pycharm problem - it resets to an empty string, which should reset to the code template. You can obtain a code template like that `step.block.options.files?.first(where: { $0.isVisible })?.text`. For reference, you can see iOS mapping [StepQuizPyCharmViewDataMapper](https://github.com/hyperskill/mobile-app/blob/0940b212b991714604be8857e5b10d6fc17e32a4/iosHyperskillApp/iosHyperskillApp/Sources/Modules/StepQuizPyCharm/StepQuizPyCharmViewDataMapper.swift).

Please also check that code reset works correctly for all code-related problems SQL, code and pycharm.

**Comment by XanderZhu** [comment]
@ivan-magda 
I fixed pycharm related problems.
But there is a reset problem for Sql (https://vyahhi.myjetbrains.com/youtrack/issue/ALTAPPS-652).
It is unrelated to this feature and will be fixed separatly.

File: androidHyperskillApp/src/main/java/org/hyperskill/app/android/debug/ui/FindStage.kt [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            keyboardOptions = KeyboardOptions(imeAction = ImeAction.Done)
+        )
+        Button(
+            onClick = onApplyClick,
+            modifier = Modifier
+                .fillMaxWidth()
+                .padding(top = Dp(20f))
+                .height(dimensionResource(id = R.dimen.action_button_height))
+        ) {
+            Text(text = stringResource(id = org.hyperskill.app.R.string.debug_menu_section_step_navigation_button_text))
```
Comment (ivan-magda) [inline androidHyperskillApp/src/main/java/org/hyperskill/app/android/debug/ui/FindStage.kt:86]:
```suggestion
            Text(text = stringResource(id = org.hyperskill.app.R.string.debug_menu_section_stage_implement_button_text))
```

**Review by ivan-magda — APPROVED** [review]
Please also include the stage implementation unsupported bottom sheet UI, see "[comment](https://vyahhi.myjetbrains.com/youtrack/issue/ALTAPPS-635/Android-Stage-implementation-unsupported-bottom-sheet#focus=Comments-74-221119.0-0)".

**Pull request merged**